cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -358,6 +359,10 @@ public ApplicationDetail getLatestAppDetail(NamespaceId namespaceId,
                                                                             appMeta.getChange()));
   }
 
+  public boolean isApplicationExists(ApplicationReference appRef) {
+    return store.getLatest(appRef.getParent(), appRef.getApplication()) != null;
```

Comment:
isn't this additional db call?

Reply:
yes it is. ApplicationID is determined from very early stage in CDAP deployment process and also as a field returned in the response, not sure if we can do that in the last DB/store step?

Reply:
umm I don't think we can even optimize the number of times we call getLatest in the deployment process - for ex: In `createApplicationVersion` we need to call getLatest again (since needs to be within the transaction)

Reply:
but what if getLatest here doesn't fetch the actual latest? Ex: Say a deployment is currently in progress which is completed after this call is triggered - then in that case we assume application doesn't exist when it actually does right? Then in that case do we overwrite the RunIds.generate(0).getId() row?

Reply:
That's a valid edge case. Even if we do this `isApplicationExists()` within the addApplication transactionRunner it still can happen. I think in this case, it's like re-deploying the same app with a new spec(like re-deploying `-SNAPSHOT` pre 6.8). Is it okay?  @albertshau @chtyim

Reply:
@sanjanasandeep what if we assert the versionID before writing it into DB:
1. If the app does not have a previous latest version, we assert it == to RunIds.generate(0).getId()
2. If it does, assert it != to RunIds.generate(0).getId()

It's bad though... And the FF is also annoying especially for the t-deprecate API that user passes in a custom version

Reply:
yeah this will handle the edge case - but right, getting hacky with FF

Reply:
checking if there is an existing row needs to be done in the same transaction as the insert, otherwise it is still open to race conditions.

Basically, Store.addApplication() needs to be modified from:

```
void addApplication(ApplicationId id, ApplicationMeta meta) throws ConflictException;
```

to:

```
ApplicationId addApplication(ApplicationReference app, ApplicationMeta meta) throws ConflictException
```

where generation of the id happens within the transaction instead of outside of it.

Reply:
The bigger problem we find is that versionID and appName is being generated far before the DB insertion transaction, and they're being used in upstream in many places... A lot of changes are needed if we determine the id from inside. Also there's deployment API that does not pass in the appName, for which we could not fetch the latest version...